pkg/pool-utils/contracts/controllers/AssetManagedLBPController.sol

```diff
@@ -0,0 +1,158 @@
+    uint256 private immutable _projectTokenIndex;
+    uint256 private immutable _reserveTokenIndex;
+    uint256 private _virtualReserveAmount;
+
+    constructor(
+        BasePoolRights memory baseRights,
+        IVault vault,
+        IERC20 reserveToken,
+        bool projectTokenFirst,
+        address manager
+    ) BasePoolController(encodePermissions(baseRights), manager) {
+        _reserveToken = reserveToken;
+        _projectTokenIndex = projectTokenFirst ? 0 : 1;
+        _reserveTokenIndex = projectTokenFirst ? 1 : 0;
+        _vault = vault;
+    }
+
+    /**
+     * @dev Call this function instead of the usual `joinPool` (init). It will borrow the necessary amount
+     * of reserve tokens using a DODO flashloan.
+     */
+    function fundPool(uint256[] memory initialBalances) external onlyManager {
+        uint256 virtualAmount = initialBalances[_getReserveTokenIndex()];
+        _virtualReserveAmount = virtualAmount;
+
+        // Set the total to the virtual balance we need
+        IVault.PoolBalanceOp[] memory ops = new IVault.PoolBalanceOp[](1);
+        ops[0] = IVault.PoolBalanceOp(IVault.PoolBalanceOpKind.UPDATE, getPoolId(), _reserveToken, virtualAmount);
+        _vault.managePoolBalance(ops);
+
+        // Pull project tokens from the manager
+        (IERC20[] memory tokens, , ) = _vault.getPoolTokens(getPoolId());
+
+        tokens[_getProjectTokenIndex()].transferFrom(
+            getManager(),
+            address(this),
+            initialBalances[_getProjectTokenIndex()]
+        );
+
+        // We are apparently joining with all tokens (initialBalances includes the reserve token)
+        // This is necessary for the invariant calculations and initial BPT minting
+        // `_onInitialize` in the pool checks if there is a managed balance: if there is, it passes
+        // zero to the Vault in amountsIn for the reserve token.
+
+        IVault.JoinPoolRequest memory request = IVault.JoinPoolRequest({
+            assets: _asIAsset(tokens),
+            maxAmountsIn: initialBalances,
+            userData: abi.encode(JoinKind.INIT, initialBalances),
+            fromInternalBalance: false
+        });
+
+        tokens[_getProjectTokenIndex()].approve(address(_vault), initialBalances[_getProjectTokenIndex()]);
+
+        // Fund the pool; pull the tokens from this contract, send BPT to the manager
+        _vault.joinPool(getPoolId(), address(this), getManager(), request);
+    }
+
+    /**
+     * @dev The initial funding results in a non-zero managed balance. When the cash balance is greater,
+     * this function can be called to return the managed funds to cash.
+     * If there is a managed balance, and the cash balance is greater, convert the managed balance to cash.
+     */
+    function restorePool() external {
+        (uint256 cash, uint256 managed, , ) = _vault.getPoolTokenInfo(getPoolId(), _reserveToken);
+
+        if (managed > 0 && cash >= _getVirtualReserveAmount()) {
```

Comment:
Is there any risk in calling this function, when the cash just exceeds or equals the _getVirtualReserveAmount. And then say like have a user sell a bunch of project tokens back to the pool shortly after. I suppose i.e. does calling restorePool() in such case result in the inability to use virtual funds and a low balance of real funds?

Reply:
That function isn't there any more. You're looking at *very* old versions; not sure how you're able to comment on non-current code